in_toto/resolver/_resolver.py

```diff
@@ -0,0 +1,61 @@
+"""Resolver interface"""
+
+from abc import ABCMeta, abstractmethod
+from pathspec import PathSpec
+import re
+
+DEFAULT_SCHEME = "default"
+RESOLVER_FOR_URI_SCHEME = {}
+
+
+def _get_scheme(uri):
+  match = re.fullmatch(r"(\w+\:)?(.*)", uri)
+  if not match:
+    raise ValueError(f"Artifact URI '{uri}' could not be parsed")
+  groups = match.groups()
+  if not groups[0]:
+    return DEFAULT_SCHEME
+  return groups[0][:-1]
+
+
+class Resolver(metaclass=ABCMeta):
+  """Interface for resolvers"""
+
+  @classmethod
+  def apply_exclude_patterns(cls, names, exclude_filter):
+    """Exclude matched patterns from passed names."""
+    included = set(names)
+
+    # Assume old way for easier testing
+    if hasattr(exclude_filter, '__iter__'):
+      exclude_filter = PathSpec.from_lines('gitwildmatch', exclude_filter)
+
+    for excluded in exclude_filter.match_files(names):
+      included.discard(excluded)
+
+    return sorted(included)
+
+  @classmethod
+  @abstractmethod
+  def resolve_uri(cls, generic_uri, **kwargs):
+    """Normalize and resolve artifact URIs"""
+    scheme = _get_scheme(generic_uri)
+
+    if scheme not in RESOLVER_FOR_URI_SCHEME:
+      raise ValueError(f"Unsupported in-toto resolver scheme '{scheme}'")
+    resolver = RESOLVER_FOR_URI_SCHEME[scheme]
+
+    return resolver.resolve_uri(generic_uri, **kwargs)
+
+  @classmethod
+  @abstractmethod
+  def hash_artifacts(cls, resolved_uri, hash_algorithms=None, **kwargs):
```

Comment:
I don't think optional arguments that differ for different resolvers are very useful here. If I need to know, which resolver I will be using, there is no need to dispatch based on the URI, right? What about something:

```python
resolver = Resolver.from_uri(uri)

if isinstance(resolver, FileResolver):
  resolver.hash_artifacts(uri, normalize_line_endings=True) 

else:
  ...
```

Reply:
I was thinking later down the line, we will allow users to write their own resolvers and place them in a directory. They might have different requirements and parameters but that could be handled as ITE-4 matures.

Reply:
Will remove `kwargs` from interface.

Reply:
```
resolver = Resolver.from_uri(uri)
```

I think this approach makes sense and is conducive to supporting custom resolvers in future, no?